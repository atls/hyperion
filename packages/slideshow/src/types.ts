import { ReactNode } from 'react'

export interface SlideshowProps {
  time?: number
  width?: string | string[]
  transition?: string
  height?: string | string[]
  controlsType?: 'dots' | 'slices'
  children?: any
}
export interface ControlsProps {
  slides?: Array<ReactNode>
  currentSlide?: number
  lastSlide?: number
  activeWidth?: number
  containerWidth?: number
}

export interface ContainerProps {
  containerWidth?: number
}
