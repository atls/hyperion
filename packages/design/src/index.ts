export * from './Sample'
export * from './Layout'
export * from './Container'
export * from './List'
export * from './Text'
export * from './Space'
