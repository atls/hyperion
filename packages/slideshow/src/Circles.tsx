import styled              from '@emotion/styled'
import React, { Fragment } from 'react'
import { ifProp }          from 'styled-tools'

const CirclesItem = styled.div<any>(
  ({ theme }) => ({
    display: 'flex',
    height: '8px',
    width: '8px',
    borderRadius: '8px',
    backgroundColor: theme.colors.lightGray,
    marginRight: '6px',
  }),
  ifProp('active', ({ theme }) => ({
    backgroundColor: theme.colors.green,
  })),
  ifProp('lastSlide', {
    marginRight: '0px',
  })
)

const CirclesContainer = styled.div<any>(({ containerWidth }) => ({
  display: 'flex',
  position: 'absolute',
  justifyContent: 'center',
  bottom: '-24px',
  zIndex: 3,
  width: `${containerWidth * 0.9}px`,
  left: `calc(50% - ${(containerWidth * 0.9) / 2}px)`,
}))

export const Circles = ({ slides, currentSlide, lastSlide, containerWidth }) => (
  <CirclesContainer containerWidth={containerWidth}>
    {slides.map((item, index) => (
      <Fragment key={`${index}as`}>
        <CirclesItem active={index === currentSlide} lastSlide={index === lastSlide} />
      </Fragment>
    ))}
  </CirclesContainer>
)
