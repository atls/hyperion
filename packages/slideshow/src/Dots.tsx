import styled              from '@emotion/styled'
import React, { Fragment } from 'react'
import { ifProp }          from 'styled-tools'

const DotsItem = styled.div<any>(
  ({ theme }) => ({
    display: 'flex',
    height: '8px',
    width: '8px',
    borderRadius: '8px',
    backgroundColor: theme.colors.lightGray,
    marginRight: '6px',
  }),
  ifProp('active', ({ theme }) => ({
    backgroundColor: theme.colors.green,
  })),
  ifProp('lastSlide', {
    marginRight: '0px',
  })
)

const DotsContainer = styled.div<any>(({ containerWidth }) => ({
  display: 'flex',
  position: 'absolute',
  justifyContent: 'center',
  bottom: '-24px',
  zIndex: 3,
  width: `${containerWidth * 0.9}px`,
  left: `calc(50% - ${(containerWidth * 0.9) / 2}px)`,
}))

export const Dots = ({ slides, currentSlide, lastSlide, containerWidth }) => (
  <DotsContainer containerWidth={containerWidth}>
    {slides.map((item, index) => (
      <Fragment key={`${item.name}`}>
        <DotsItem active={index === currentSlide} lastSlide={index === lastSlide} />
      </Fragment>
    ))}
  </DotsContainer>
)
