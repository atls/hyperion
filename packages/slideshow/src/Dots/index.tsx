import React, { FC, Fragment } from 'react'

import { unique }              from '@atlantis-lab/utils'

import { ControlsProps }       from '../types'
import { Container }           from './Container'
import { Dot }                 from './Dot'

export const Dots: FC<ControlsProps> = ({ slides, currentSlide, lastSlide, containerWidth }) => {
  return (
    <Container containerWidth={containerWidth}>
      {slides.map((item, index) => (
        <Fragment key={`${unique()}`}>
          <Dot active={index === currentSlide} lastSlide={index === lastSlide} />
        </Fragment>
      ))}
    </Container>
  )
}
