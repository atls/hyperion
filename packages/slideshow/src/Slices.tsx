import styled              from '@emotion/styled'
import React, { Fragment } from 'react'
import { ifProp }          from 'styled-tools'

const StyledSlice = styled.div<any>(
  {
    display: 'flex',
    height: '2px',
    backgroundColor: 'white',
    opacity: '0.5',
    marginRight: '6px',
  },
  ifProp('active', {
    opacity: '1',
  }),
  ifProp('lastSlide', {
    marginRight: '0px',
  })
)

const SlicesContainer = styled.div<any>(({ containerWidth }) => ({
  display: 'flex',
  position: 'absolute',
  justifyContent: 'center',
  bottom: '30px',
  zIndex: 3,
  width: `${containerWidth * 0.9}px`,
  left: `calc(50% - ${(containerWidth * 0.9) / 2}px)`,
}))

export const Slices = ({ slides, currentSlide, lastSlide, containerWidth }) => (
  <SlicesContainer containerWidth={containerWidth}>
    {slides.map((item, index) => (
      <Fragment key={`${item.name}`}>
        <StyledSlice
          style={{ flexBasis: `${containerWidth / slides.length - 6}px` }}
          active={index === currentSlide}
          lastSlide={index === lastSlide}
        />
      </Fragment>
    ))}
  </SlicesContainer>
)
