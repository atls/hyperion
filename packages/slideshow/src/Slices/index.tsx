import React, { FC }     from 'react'

import { Box }           from '@atlantis-lab/layout'

import { ControlsProps } from '../types'
import { Container }     from './Container'
import { Slice }         from './Slice'

/* eslint-disable */
export const Slices: FC<ControlsProps> = ({
  slides,
  activeWidth,
  currentSlide,
  lastSlide,
  containerWidth,
}) => (
  <Container containerWidth={containerWidth}>
    {slides.map((item, index) => (
      <Box
        key={`slice-${index}`}
        position='relative'
        flexBasis={`${containerWidth / slides.length - 6}px`}
        height='2px'
        backgroundColor='rgba(255, 255, 255, 0.5)'
        overflow='hidden'
        marginRight={lastSlide === index ? '0px' : '6px'}
      >
        <Slice activeWidth={index === currentSlide ? activeWidth : '0px'} />
      </Box>
    ))}
  </Container>
)
/* eslint-enable */
