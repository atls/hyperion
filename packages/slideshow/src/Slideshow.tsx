import styled                                     from '@emotion/styled'
import React, { FC, useEffect, useRef, useState } from 'react'
import { Swipeable }                              from 'react-swipeable'
import { layout }                                 from 'styled-system'

import { contentWidth }                           from '@atlantis-lab/utils'

import { Dots }                                   from './Dots'
import { Slices }                                 from './Slices'
import { SlideshowProps }                         from './types'

type SlideshowComponentProps = {
  children: any
  width?: string
  height?: string
  transition?: string
  time?: number
  controlsType?: 'slices' | 'dots'
}

const Container = styled.div<any>(
  {
    display: 'flex',
    position: 'relative',
    touchAction: 'pan-y',
  },
  layout
)

const StyledSlide = styled.div<any>(({ opacity, transition, visibility }) => ({
  position: 'absolute',
  transition,
  opacity,
  visibility,
}))

export const Slideshow: FC<SlideshowProps> = ({
  children,
  width,
  height,
  transition,
  time = 10000,
  controlsType = 'slices',
}: SlideshowComponentProps) => {
  const [slide, setSlide] = useState(0)
  const [activeWidth, setActiveWidth] = useState(0)
  const [stop, setStop] = useState(false)
  const [containerWidth, setContainerWidth] = useState(null)
  const containerNode = useRef(null)

  const stopSlideshow = () => {
    setActiveWidth(100)
    setStop(true)
  }

  const swiped = data => {
    if (data.dir !== 'Up' && data.dir !== 'Down') {
      if (data.dir === 'Left') {
        if (slide < children.length - 1) {
          setSlide(slide + 1)
        } else {
          setSlide(0)
        }
      } else if (slide > 0) {
        setSlide(slide - 1)
      } else {
        setSlide(children.length - 1)
      }
      stopSlideshow()
    }
  }

  useEffect(() => {
    setContainerWidth(contentWidth(containerNode.current))
  }, [children.length])

  useEffect(() => {
    if (activeWidth > 110) {
      if (slide < children.length - 1) {
        setSlide(slide + 1)
        setActiveWidth(0)
      } else {
        setSlide(0)
        setActiveWidth(0)
      }
    }
  }, [children.length, activeWidth])

  useEffect(() => {
    let timeout = null
    if (!stop) {
      timeout = setTimeout(() => {
        setActiveWidth(activeWidth + 2)
      }, time / 50)
    }
    return () => clearTimeout(timeout)
  }, [activeWidth, stop])
  /* eslint-disable */
  return (
    <Container ref={containerNode} onMouseEnter={stopSlideshow} width={width} height={height}>
      <Swipeable onSwiped={(data) => swiped(data)} trackMouse trackTouch delta={30}>
        {children.map((item, index) => (
          <StyledSlide
            transition={transition}
            key={`slide-${index}`}
            opacity={slide === index ? 1 : 0}
            visibility={slide === index ? 'visible' : 'hidden'}
          >
            {item}
          </StyledSlide>
        ))}
      </Swipeable>
      {controlsType === 'dots' && (
        <Dots
          slides={children}
          currentSlide={slide}
          containerWidth={containerWidth}
          lastSlide={children.length - 1}
        />
      )}
      {controlsType === 'slices' && (
        <Slices
          slides={children}
          currentSlide={slide}
          activeWidth={activeWidth}
          containerWidth={containerWidth}
          lastSlide={children.length - 1}
        />
      )}
    </Container>
  )
  /* eslint-enable */
}
