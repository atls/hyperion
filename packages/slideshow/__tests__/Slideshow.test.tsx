import React         from 'react'
import renderer      from 'react-test-renderer'

import { Slideshow } from '../src/index'

it('Slideshow should render correctly', () => {
  const tree = renderer
    .create(
      <Slideshow transition='.5s'>
        <h1>Slide 1</h1>
        <h1>Slide 2</h1>
        <h1>Slide 3</h1>
      </Slideshow>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
