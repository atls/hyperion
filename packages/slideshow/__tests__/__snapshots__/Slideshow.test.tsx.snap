// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slideshow should render correctly 1`] = `
.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  touch-action: pan-y;
}

.emotion-0 {
  position: absolute;
  -webkit-transition: .5s;
  transition: .5s;
  opacity: 1;
  visibility: visible;
}

.emotion-1 {
  position: absolute;
  -webkit-transition: .5s;
  transition: .5s;
  opacity: 0;
  visibility: hidden;
}

.emotion-3 {
  position: relative;
  visibility: hidden;
  opacity: 0;
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  bottom: 30px;
  z-index: 3;
  width: 0px;
  left: calc(50% - 0px);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2px;
  position: absolute;
  background-color: white;
  -webkit-transition: 0.35s;
  transition: 0.35s;
  width: 0%;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2px;
  position: absolute;
  background-color: white;
  -webkit-transition: 0.35s;
  transition: 0.35s;
  width: 0px%;
}

.emotion-5 {
  box-sizing: border-box;
  height: 2px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 6px;
  background-color: rgba(255, 255, 255, 0.5);
  -webkit-flex-basis: -6px;
  -ms-flex-preferred-size: -6px;
  flex-basis: -6px;
  position: relative;
}

.emotion-9 {
  box-sizing: border-box;
  height: 2px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0px;
  background-color: rgba(255, 255, 255, 0.5);
  -webkit-flex-basis: -6px;
  -ms-flex-preferred-size: -6px;
  flex-basis: -6px;
  position: relative;
}

<div
  className="emotion-11"
  onMouseEnter={[Function]}
>
  <div
    onMouseDown={[Function]}
  >
    <div
      className="emotion-0"
      opacity={1}
      visibility="visible"
    >
      <h1>
        Slide 1
      </h1>
    </div>
    <div
      className="emotion-1"
      opacity={0}
      visibility="hidden"
    >
      <h1>
        Slide 2
      </h1>
    </div>
    <div
      className="emotion-1"
      opacity={0}
      visibility="hidden"
    >
      <h1>
        Slide 3
      </h1>
    </div>
  </div>
  <div
    className="emotion-3"
  >
    <div />
  </div>
  <div
    className="emotion-10"
  >
    <div
      className="emotion-5"
      display="flex"
      height="2px"
      overflow="hidden"
    >
      <div
        className="emotion-4"
      />
    </div>
    <div
      className="emotion-5"
      display="flex"
      height="2px"
      overflow="hidden"
    >
      <div
        className="emotion-6"
      />
    </div>
    <div
      className="emotion-9"
      display="flex"
      height="2px"
      overflow="hidden"
    >
      <div
        className="emotion-6"
      />
    </div>
  </div>
</div>
`;
